Framework/Algorithms/src/CrossCorrelate.cpp

```diff
@@ -96,147 +102,159 @@ void CrossCorrelate::init() {
+  m_progress = std::make_unique<Progress>(this, 0.0, 1.0, numSpectra);
   PARALLEL_FOR_IF(Kernel::threadSafe(*inputWS, *out))
-  for (int i = 0; i < nspecs; ++i) // Now loop on all spectra
+  for (int currentSpecIndex = 0; currentSpecIndex < numSpectra; ++currentSpecIndex) // Now loop on all spectra
   {
     PARALLEL_START_INTERUPT_REGION
-    size_t wsIndex = indexes[i]; // Get the ws index from the table
+    size_t wsIndex = indexes[currentSpecIndex]; // Get the ws index from the table
     // Copy spectra info from input Workspace
-    out->getSpectrum(i).copyInfoFrom(inputWS->getSpectrum(wsIndex));
-
-    out->setSharedX(i, out->sharedX(0));
-
-    // Get temp references
-    const auto &iX = inputWS->x(wsIndex);
-    const auto &iY = inputWS->y(wsIndex);
-    const auto &iE = inputWS->e(wsIndex);
-    // Copy Y,E data of spec(i) to temp vector
-    // Now rebin on the grid of reference spectrum
-    std::vector<double> tempY(nY);
-    std::vector<double> tempE(nY);
-    VectorHelper::rebin(iX.rawData(), iY.rawData(), iE.rawData(), refX, tempY, tempE, is_distrib);
+    out->getSpectrum(currentSpecIndex).copyInfoFrom(inputWS->getSpectrum(wsIndex));
+
+    out->setSharedX(currentSpecIndex, out->sharedX(0));
+
+    // Get temp referenceSpectras
+    const auto &inputXVector = inputWS->x(wsIndex);
+    const auto &inputYVector = inputWS->y(wsIndex);
+    const auto &inputEVector = inputWS->e(wsIndex);
+    // Copy Y,E data of spec(currentSpecIndex) to temp vector
+    // Now rebin on the grid of referenceSpectra
+    std::vector<double> tempY(numReferenceY);
+    std::vector<double> tempE(numReferenceY);
+    VectorHelper::rebin(inputXVector.rawData(), inputYVector.rawData(), inputEVector.rawData(), referenceXVector, tempY,
+                        tempE, isDistribution);
     const auto tempVar = subtractMean(tempY, tempE);
 
     // Calculate the normalisation constant
     const double tempNorm = 1.0 / sqrt(tempVar.y);
     const double tempNormE = 0.5 * pow(tempNorm, 3) * sqrt(tempVar.e);
-    const double normalisation = refNorm * tempNorm;
-    const double normalisationE2 = pow((refNorm * tempNormE), 2) + pow((tempNorm * refNormE), 2);
-    // Get reference to the ouput spectrum
-    auto &outY = out->mutableY(i);
-    auto &outE = out->mutableE(i);
-
-    for (int k = -nY + 2; k <= nY - 2; ++k) {
-      const int kp = abs(k);
+    const double normalisation = referenceNorm * tempNorm;
+    const double normalisationE2 = pow((referenceNorm * tempNormE), 2) + pow((tempNorm * referenceNormE), 2);
+    // Get referenceSpectr to the ouput spectrum
+    auto &outY = out->mutableY(currentSpecIndex);
+    auto &outE = out->mutableE(currentSpecIndex);
+
+    // max the shift
+    int shiftCorrection = 0;
+    if (maxDSpaceShift != EMPTY_DBL()) {
+      // convert dspacing to bins, where maxDSpaceShift is at least 0.1
+      const auto maxBins = std::max(0.0 + maxDSpaceShift, 0.1) / (rangeEnd - rangeStart);
```

Comment:
I don't think this is correct, I think it should be something like
```
const auto maxBins = std::max(0.0 + maxDSpaceShift*2, 0.1) / (xmax - xmin) * (rangeEnd - rangeStart);
```

Your PR test instructions have `MaxDSpaceShift=100` which doesn't make sense if the entire data range is only 5.

Reply:
I made the suggested changes but I'm not sure the output is correct, as for integer values  of `maxDSpaceShift` above 2 the graph is just a flat line.  I imagine its a flatline because I am not upper bounding the shift correction, and its going out of range or something(now that the maxBins value is being multiplied by the range size), but does that sound like its expected behavior or should `maxDSpaceShift` be able to go higher?

100 is the starting value, the instructions say to adjust it during testing to see if the parameter change properly bounds it.  It was explained to me that this algo was used such that the `maxDSpaceShift` parameter's value would get lowered until the result of the graph changed, thus finding the correct bound.  So I started the script on a high number.

Reply:
Yeah it looks like the shift correction should not have that outside `abs` and that the minimum should be 0, so something like
```c++
shiftCorrection = std::max(0.0, abs((-numReferenceY + 2) - (numReferenceY - 2)) - maxBins) / 2;
```

Reply:
I am testing this with the following more complicated example

```python
myFunc = "name=LinearBackground, A0=10;name=Gaussian, PeakCentre=2, Height=10, Sigma=0.1;name=Gaussian, PeakCentre=10, Height=10, Sigma=0.1;name=Gaussian, PeakCentre=12, Height=10, Sigma=0.1"
ws = CreateSampleWorkspace("Histogram","User Defined", myFunc, BankPixelWidth=1, XUnit='dSpacing', XMax=20, BinWidth=0.01)
ws = ScaleX(InputWorkspace='ws', Factor=0.25, Operation='Add', IndexMin=1, IndexMax=1)
OutputWorkspace = CrossCorrelate(InputWorkspace='ws', WorkspaceIndexMax=1, XMin=2, XMax=18, MaxDSpaceShift=5)
```

The output with these fixes looks correct, here I have `MaxDSpaceShift=5` where the bin size is 0.01, so that should be a max of +-500 bins which is what I see.
![OutputWorkspace-1](https://user-images.githubusercontent.com/5595210/121067163-b759bc80-c798-11eb-903a-9c6bb0864121.png)